File: crates/suspect-codegen/src/ruby_sdk/streams.rb
# frozen_string_literal: true
module __NAMESPACE__
  # Backpressured, source-validated native Enumerator. Use #each (which ensures
  # cleanup) or close explicitly after #next. The total deadline remains active
  # even while the consumer is paused. No sentinel or reconnect loop exists.
  class ItemStream < Enumerator
    attr_reader :source
    def initialize(lease, descriptor, context, values: nil)
      @lease, @descriptor, @context = lease, descriptor, context
      @source = Internal::Wire.source(descriptor.fetch('source'))
      @closed = false
      super() do |out|
        begin
          if values
            values.each { |value| out << value }
          else
            parser = Internal::ItemParser.new(descriptor, context)
            parser.each(lease) { |value| out << value }
          end
        ensure
          close
        end
      end
    end
    def close
      return nil if @closed
      @closed = true
      @lease&.close
      nil
    end
    def closed? = @closed
    def next
      @context&.check!
      raise StopIteration, 'stream is closed' if @closed
      super
    end
    def inspect = "#<#{self.class.name} closed=#{@closed}>"
    # @api private
    def self.completed(values, source)
      new(nil, {'source' => source}, nil, values: values)
    end
  end

  # @api private
  module Internal
    class LeaseClosed < Exception; end
    # Block-based transport lifetime is retained on one worker. Only one bounded
    # chunk is handed to a waiting consumer; reads are demand-driven.
    class ExchangeLease
      attr_reader :status, :headers
      def initialize(transport, request, context)
        @context, @commands, @chunks, @ready = context, Queue.new, SizedQueue.new(1), Queue.new
        @closed, @error, @done = false, nil, false
        @worker = Thread.new do
          Thread.current.report_on_exception = false
          yielded, ready = false, false
          begin
            context.run do
              transport.exchange(request: request, context: context) do |response|
                begin
                  raise TransportError, 'transport yielded multiple responses' if yielded
                  yielded = true
                  @ready << [response.status, response.headers]
                  ready = true
                  @commands.pop
                  response.each_chunk do |chunk|
                    context.check!
                    raise TransportError, 'body chunks must be Strings' unless chunk.instance_of?(String)
                    if chunk.empty?
                      @chunks << ''.b
                      @commands.pop
                    else
                      offset = 0
                      while offset < chunk.bytesize
                        @chunks << chunk.byteslice(offset, 16 * 1024).b.freeze
                        offset += 16 * 1024
                        @commands.pop
                      end
                    end
                  end
                ensure
                  primary = $!
                  begin response.close
                  rescue StandardError
                    raise unless primary
                  end
                end
              end
              raise TransportError, 'transport did not yield a response' unless yielded
            end
          rescue LeaseClosed
            nil
          rescue Exception => error
            @error = error
          ensure
            @done = true
            @ready << nil unless ready
            @chunks.close
          end
        end
        first = @ready.pop
        raise_error if @error
        raise TransportError, 'transport did not produce metadata' unless first
        @status, @headers = first
      rescue Exception
        close
        raise
      end
      def each_chunk
        loop do
          @context.check!
          raise_error if @error
          break if @closed
          @commands << true
          chunk = @chunks.pop
          raise_error if @error
          break unless chunk
          yield chunk
        end
        nil
      end
      def close
        return nil if @closed
        @closed = true
        @worker.raise(LeaseClosed) if @worker&.alive? && !@done
        @worker&.join
        nil
      end
      def closed? = @closed
      def raise_error
        error = @error
        raise error if error.is_a?(SdkError) || !error.is_a?(StandardError)
        if error.is_a?(::Net::OpenTimeout) || error.is_a?(::Net::ReadTimeout) || error.is_a?(::Net::WriteTimeout)
          raise TimeoutError.new('stream transport timed out', source: @context.source, operation_id: @context.operation_id), cause: error
        end
        raise TransportError.new('stream transport failed', source: @context.source, operation_id: @context.operation_id), cause: error
      end
    end

    class LimitedChunks
      attr_reader :capture, :truncated, :bytes
      def initialize(response, context, limit, capture_limit, metadata)
        @response, @context, @limit, @capture_limit, @metadata = response, context, limit, capture_limit, metadata
        @capture, @bytes, @count, @truncated = +''.b, 0, 0, false
      end
      def each_chunk
        @response.each_chunk do |chunk|
          @context.check!
          raise TransportError.new('body chunks must be Strings', kind: :transport, **details) unless chunk.instance_of?(String)
          @count += 1; @bytes += chunk.bytesize
          left = @capture_limit - @capture.bytesize
          @capture << chunk.byteslice(0, left).b if left.positive?
          @truncated ||= chunk.bytesize > left
          if @count > POLICY[:max_response_chunks] || @bytes > @limit
            @truncated = true
            raise ResourceLimitError.new('response byte/chunk ceiling exceeded', kind: :resource_limit, **details)
          end
          yield chunk
        end
        nil
      end
      def details = @metadata.merge(capture: @capture.dup.freeze, truncated: @truncated)
      def close = @response.close
      def closed? = @response.closed?
    end

    class ItemParser
      def initialize(descriptor, context)
        @descriptor, @context = descriptor, context
        @max = Wire.integer(descriptor['max_item_bytes'])
        @codec, @items = CodecSession.new, 0
        @source = Wire.source(descriptor['source'])
      end
      def emit(value)
        @context.check!
        @items += 1
        raise ResourceLimitError.new('stream item count ceiling exceeded', kind: :resource_limit, source: @source) if @items > POLICY[:max_stream_items]
        index = Wire.codec_index(@descriptor['item_codec'])
        yield @codec.decode(index, value)
      rescue CodecError, JsonError => error
        raise ResponseError.new('stream item does not satisfy its source codec', kind: :response_decoding, source: error.source || @source, operation_id: @context.operation_id), cause: error
      end
      def each(chunks)
        buffer = +''.b; first = true; skip_lf = false
        sse = @descriptor['framing'] == 'server-sent-events'
        data, event, id, retry_value, event_bytes = [], nil, nil, nil, 0
        process = lambda do |raw|
          text = raw.dup.force_encoding(Encoding::UTF_8)
          if sse
            # WHATWG UTF-8 decoding replaces invalid sequences for event streams.
            text = text.scrub
            text = text.delete_prefix("\uFEFF") if first
            first = false
            event_bytes += raw.bytesize + 1
            raise ResourceLimitError, 'SSE event byte ceiling exceeded' if event_bytes > @max
            if text.empty?
              unless data.empty?
                value = {'data' => data.join("\n")}
                value['event'] = event unless event.nil? || event.empty?
                value['id'] = id unless id.nil?
                value['retry'] = retry_value unless retry_value.nil?
                emit(value) { |v| yield v }
              end
              data, event, event_bytes = [], nil, 0
            elsif !text.start_with?(':')
              key, value = text.split(':', 2); value ||= ''; value = value.delete_prefix(' ')
              case key
              when 'data' then data << value
              when 'event' then event = value
              when 'id' then id = value unless value.include?("\0")
              when 'retry'
                if /\A[0-9]+\z/n.match?(value.b)
                  # HTML retry is decimal digits, not Ruby's padded/octal syntax.
                  digits = value.sub(/\A0+/, ''); digits = '0' if digits.empty?
                  retry_value = JsonNumber.new(digits)
                end
              end
            end
          else
            text = text.delete_suffix("\r")
            raise ResponseError.new('blank JSON line', kind: :response_decoding, source: @source, operation_id: @context.operation_id) if text.empty?
            emit(Json.parse(text, max_bytes: @max)) { |v| yield v }
          end
        end
        chunks.each_chunk do |chunk|
          chunk.each_byte do |byte|
            if sse && skip_lf
              skip_lf = false
              next if byte == 10
            end
            if byte == 10 || sse && byte == 13
              process.call(buffer)
              buffer = +''.b
              skip_lf = sse && byte == 13
            else
              buffer << byte
              raise ResourceLimitError.new('stream line/item byte ceiling exceeded', kind: :resource_limit, source: @source) if buffer.bytesize > @max
            end
          end
        end
        # SSE EOF discards an unfinished event; JSON lines permit a final record
        # without a trailing LF. No [DONE] or record-separator shortcut exists.
        process.call(buffer) unless sse || buffer.empty?
      rescue JsonError, CodecError => error
        raise ResponseError.new('invalid stream framing/item', kind: :response_decoding, source: @source, operation_id: @context.operation_id), cause: error
      end
    end
  end
end

File: crates/suspect-codegen/src/python_http/streams.py
"""Caller-owned, context-managed item streams with bounded incremental framing."""
from __future__ import annotations
from collections.abc import AsyncIterator, Awaitable, Callable, Iterator
from typing import Generic, TypeVar, cast
from . import json_runtime as J
from .codec_runtime import CodecError, ModelCodec
from ._types import Source, SdkError
from ._wire import text_bytes

T = TypeVar('T')
Failure = Callable[[str, BaseException | None], SdkError]


class Framer:
    def __init__(self, framing: str, limit: int, source: Source) -> None:
        self.framing, self.limit, self.source = framing, limit, source
        self.line = bytearray()
        self.frame_bytes = 0
        self.skip_lf = False
        self.first = True
        self.data: list[str] = []
        self.event: str | None = None
        self.last_id: str | None = None
        self.retry: int | None = None

    def feed(self, chunk: bytes) -> Iterator[J.JsonValue]:
        for byte in chunk:
            if self.framing == 'server-sent-events' and self.skip_lf:
                self.skip_lf = False
                if byte == 10:
                    continue
            self.frame_bytes += 1
            if self.frame_bytes > self.limit:
                raise SdkError('resource-limit', self.source, code='http-stream-item-limit')
            newline = byte == 10 or self.framing == 'server-sent-events' and byte == 13
            if not newline:
                self.line.append(byte)
                continue
            self.skip_lf = self.framing == 'server-sent-events' and byte == 13
            line = bytes(self.line)
            self.line.clear()
            yield from self.accept(line)

    def accept(self, raw: bytes) -> Iterator[J.JsonValue]:
        if self.framing == 'json-lines':
            if raw.endswith(b'\r'):
                raw = raw[:-1]
            self.frame_bytes = 0
            # A blank line is not a JSON value. Final LF is optional; no RS or
            # sentinel convention is introduced by this framing.
            yield J.parse_json(raw)
            return
        # HTML event-stream uses UTF-8 replacement decoding and ignores one
        # leading BOM. Line bytes are retained across arbitrary chunk splits.
        text = raw.decode('utf-8', 'replace')
        if self.first:
            self.first = False
            if text.startswith('\ufeff'):
                text = text[1:]
        if not text:
            self.frame_bytes = 0
            if self.data:
                event: dict[str, J.JsonValue] = {'data': '\n'.join(self.data)}
                if self.last_id is not None:
                    event['id'] = self.last_id
                if self.event is not None:
                    event['event'] = self.event or 'message'
                if self.retry is not None:
                    event['retry'] = self.retry
                yield event
            self.data.clear()
            self.event = None
            return
        if text.startswith(':'):
            return
        name, separator, value = text.partition(':')
        if not separator:
            value = ''
        elif value.startswith(' '):
            value = value[1:]
        if name == 'data':
            self.data.append(value)
        elif name == 'event':
            self.event = value
        elif name == 'id' and '\0' not in value:
            self.last_id = value
        elif name == 'retry' and value and all('0' <= c <= '9' for c in value):
            self.retry = J.JsonNumber(value.lstrip('0') or '0').to_int()

    def finish(self) -> Iterator[J.JsonValue]:
        if self.framing == 'json-lines' and self.line:
            yield from self.accept(bytes(self.line))
        # HTML discards an unfinished event at EOF, even after a data line.
        self.line.clear()
        self.data.clear()


class SyncStream(Generic[T], Iterator[T]):
    """Validate one item per pull. Use `with` or close() when stopping early."""
    def __init__(self, chunks: Iterator[bytes], codec: ModelCodec[T], framing: str,
                 limit: int, source: Source, close: Callable[[BaseException | None], None], failure: Failure,
                 owner_closed: Callable[[], bool] | None = None) -> None:
        self._chunks, self._codec = chunks, codec
        self._framer = Framer(framing, limit, source)
        self._close, self._failure = close, failure
        self._items: Iterator[J.JsonValue] = iter(())
        self._finished = False
        self._closed = False
        self._owner_closed = owner_closed

    def __iter__(self) -> SyncStream[T]:
        return self

    def __next__(self) -> T:
        if self._closed or self._owner_closed is not None and self._owner_closed():
            self._closed = True
            raise StopIteration
        try:
            while True:
                try:
                    value = next(self._items)
                except StopIteration:
                    if self._finished:
                        self.close()
                        raise
                    try:
                        chunk = next(self._chunks)
                    except StopIteration:
                        self._finished = True
                        self._items = self._framer.finish()
                    else:
                        self._items = self._framer.feed(chunk)
                    continue
                return self._codec.decode_value(value)
        except StopIteration:
            raise
        except BaseException as error:
            self._closed = True
            self._close(error)
            if isinstance(error, (CodecError, J.JsonError)):
                kind = 'resource-limit' if error.kind in ('resource', J.RESOURCE_LIMIT) else 'response-decoding'
                raise self._failure(kind, error) from None
            if isinstance(error, SdkError) and error.status is None:
                failure = self._failure(error.kind, error)
                failure.code = error.code
                raise failure from None
            raise

    def close(self) -> None:
        if not self._closed:
            self._closed = True
            self._close(None)

    def __enter__(self) -> SyncStream[T]:
        return self

    def __exit__(self, kind: object, error: BaseException | None, traceback: object) -> None:
        if not self._closed:
            self._closed = True
            self._close(error)


class AsyncStream(Generic[T], AsyncIterator[T]):
    """Await one item per pull. Async context exit closes an early-stopped body."""
    def __init__(self, chunks: AsyncIterator[bytes], codec: ModelCodec[T], framing: str,
                 limit: int, source: Source, close: Callable[[BaseException | None], Awaitable[None]], failure: Failure,
                 owner_closed: Callable[[], bool] | None = None) -> None:
        self._chunks, self._codec = chunks, codec
        self._framer = Framer(framing, limit, source)
        self._close, self._failure = close, failure
        self._items: Iterator[J.JsonValue] = iter(())
        self._finished = False
        self._closed = False
        self._owner_closed = owner_closed

    def __aiter__(self) -> AsyncStream[T]:
        return self

    async def __anext__(self) -> T:
        if self._closed or self._owner_closed is not None and self._owner_closed():
            self._closed = True
            raise StopAsyncIteration
        try:
            while True:
                try:
                    value = next(self._items)
                except StopIteration:
                    if self._finished:
                        await self.aclose()
                        raise StopAsyncIteration from None
                    try:
                        chunk = await self._chunks.__anext__()
                    except StopAsyncIteration:
                        self._finished = True
                        self._items = self._framer.finish()
                    else:
                        self._items = self._framer.feed(chunk)
                    continue
                return self._codec.decode_value(value)
        except StopAsyncIteration:
            raise
        except BaseException as error:
            self._closed = True
            await self._close(error)
            if isinstance(error, (CodecError, J.JsonError)):
                kind = 'resource-limit' if error.kind in ('resource', J.RESOURCE_LIMIT) else 'response-decoding'
                raise self._failure(kind, error) from None
            if isinstance(error, SdkError) and error.status is None:
                failure = self._failure(error.kind, error)
                failure.code = error.code
                raise failure from None
            raise

    async def aclose(self) -> None:
        if not self._closed:
            self._closed = True
            await self._close(None)

    async def __aenter__(self) -> AsyncStream[T]:
        return self

    async def __aexit__(self, kind: object, error: BaseException | None, traceback: object) -> None:
        if not self._closed:
            self._closed = True
            await self._close(error)


def encode(value: J.JsonValue, framing: str, maximum: int, source: Source) -> bytes:
    if framing == 'json-lines':
        return text_bytes(J.stringify_json(value) + '\n', maximum, source)
    if type(value) is not dict:
        raise SdkError('request-representation', source, code='http-sse-envelope')
    event = value
    if 'data' not in event or any(key not in ('data', 'event', 'id', 'retry') for key in event):
        raise SdkError('request-representation', source, code='http-sse-envelope')
    pieces: list[str] = []
    for key in ('id', 'event', 'retry', 'data'):
        if key not in event:
            continue
        item = event[key]
        if key == 'retry':
            if type(item) is J.JsonNumber:
                item = item.to_int()
            if type(item) is not int or item < 0:
                raise SdkError('request-representation', source, code='http-sse-retry')
            pieces.append('retry:' + J.stringify_json(item) + '\n')
        else:
            if type(item) is not str or '\r' in item or key != 'data' and ('\n' in item or '\0' in item):
                raise SdkError('request-representation', source, code='http-sse-field')
            # One separator space is removed by the HTML parser; adding it
            # preserves any leading spaces that belong to the field value.
            pieces.extend(key + ': ' + line + '\n' for line in item.split('\n'))
    return text_bytes(''.join(pieces) + '\n', maximum, source)
